packages/@aws-cdk/aws-lambda-python/lib/bundling.ts

```diff
@@ -16,39 +26,65 @@ export interface BundlingOptions {
    * The runtime of the lambda function
    */
   readonly runtime: lambda.Runtime;
+
+  /**
+   * Output path suffix ('python' for a layer, '.' otherwise)
+   */
+  readonly outputPathSuffix: string;
 }
 
 /**
  * Produce bundled Lambda asset code
  */
 export function bundle(options: BundlingOptions): lambda.AssetCode {
-  // Bundling image derived from runtime bundling image (AWS SAM docker image)
-  const image = cdk.BundlingDockerImage.fromAsset(__dirname, {
-    buildArgs: {
-      IMAGE: options.runtime.bundlingDockerImage.image,
-    },
-  });
-
-  let installer = options.runtime === lambda.Runtime.PYTHON_2_7 ? Installer.PIP : Installer.PIP3;
+  const { entry, runtime, outputPathSuffix: pathSuffix } = options;
 
-  let hasRequirements = fs.existsSync(path.join(options.entry, 'requirements.txt'));
+  const hasDeps = hasDependencies(entry);
 
-  let depsCommand = chain([
-    hasRequirements ? `${installer} install -r requirements.txt -t ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}` : '',
-    `rsync -r . ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}`,
+  const depsCommand = chain([
+    hasDeps ? `rsync -r ${BUNDLER_DEPENDENCIES_CACHE}/. ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}/${pathSuffix}` : '',
+    `rsync -r . ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}/${pathSuffix}`,
   ]);
 
-  return lambda.Code.fromAsset(options.entry, {
+  // Determine which dockerfile to use. When dependencies are present, we use a
+  // Dockerfile that can create a cacheable layer. We can't use this Dockerfile
+  // if there aren't dependencies or the Dockerfile will complain about missing
+  // sources.
+  const dockerfile = hasDeps
+    ? 'Dockerfile.dependencies'
+    : 'Dockerfile';
+
+  const image = cdk.BundlingDockerImage.fromAsset(entry, {
+    buildArgs: {
+      IMAGE: runtime.bundlingDockerImage.image,
+    },
+    file: path.join(__dirname, dockerfile),
+  });
+
+  return lambda.Code.fromAsset(entry, {
+    assetHashType: cdk.AssetHashType.BUNDLE,
+    exclude: DEPENDENCY_EXCLUDES,
     bundling: {
       image,
       command: ['bash', '-c', depsCommand],
     },
   });
 }
 
-enum Installer {
-  PIP = 'pip',
-  PIP3 = 'pip3',
+/**
+ * Checks to see if the `entry` directory contains a type of dependency that
+ * we know how to install.
+ */
+export function hasDependencies(entry: string): boolean {
+  if (fs.existsSync(path.join(entry, 'Pipfile'))) {
+    return true;
+  }
+
+  if (fs.existsSync(path.join(entry, 'requirements.txt'))) {
+    return true;
+  }
+
+  return false;
```

Comment:
*personal preference/style

```suggestion
  const pipfileExists = fs.existsSync(path.join(entry, 'Pipfile'));
  const requirementsFileExists = fs.existsSync(path.join(entry, 'requirements.txt'));
  
  return pipfileExists || requirementsFileExists
```